orangecontrib/spectroscopy/io/perkinelmer.py

```diff
@@ -0,0 +1,471 @@
+
+                else:
+                    spectrum = decoded
+
+            reader.start += block_size
+
+        wavenumbers = np.linspace(info['min_wavelength'],
+                                  info['max_wavelength'],
+                                  info['n_points'])
+
+        datavals = np.array(spectrum)[None, ...]
+
+        domain = Domain([], None)
+        meta_data = Table.from_numpy(domain,
+                                     X=np.zeros((len(datavals), 0)))
+
+        meta_data.attributes = info
+
+        return wavenumbers, datavals, meta_data
+
+    def read_fsm(self):
+        f = open(self.filename, "rb")
+        data = f.read()
+        f.close()
+
+        reader = BlockReader(data)
+
+        meta = {
+            "signature": reader.read(4, format="utf-8"),
+            "description": reader.read(40, format="utf-8"),
+        }
+
+        decoders = {
+            5100: PerkinElmer.decode5100,
+            5104: PerkinElmer.decode5104,
+            5105: PerkinElmer.decode5105,
+        }
+
+        spectrum = []
+
+        while not reader.atEnd(6):
+            block_id, block_size = reader.read(6,
+                                               format="<Hi",
+                                               expect_tuple=True)
+
+            decoded = decoders[block_id](reader.read(block_size))
+
+            if isinstance(decoded, dict):
+                meta.update(decoded)
+            else:
+                spectrum.append(decoded)
+
+        wavenumbers = np.arange(meta['z_start'],
+                                meta['z_end'] + meta['z_delta'],
+                                meta['z_delta'])
+
+        datavals = np.squeeze(spectrum)
+
+        domain = Domain([], None)
+        meta_data = Table.from_numpy(domain,
+                                     X=np.zeros((len(datavals), 0)))
+
+        meta_data.attributes = meta
+
+        return datavals, wavenumbers, meta_data
+
+    def read_spectra(self):
+        if self.filename[-2:] == "sp":
+            return self.read_sp()
+
+        elif self.filename[-3:] == "fsm":
+            intensities, wn, m = self.read_fsm()
+
+            # TODO position calculation should be done by a helper function that accepts units as well
+            xpositions = m.attributes['x_init'] + m.attributes['x_delta'] * np.arange(m.attributes['n_x'])
+            ypositions = m.attributes['y_init'] + m.attributes['y_delta'] * np.arange(m.attributes['n_y'])
+
+            # TODO this is not nice - we need new helpers to build the coordinates
+            y_loc = np.repeat(np.arange(m.attributes['n_y']), m.attributes['n_x'])
+            x_loc = np.tile(np.arange(m.attributes['n_x']), m.attributes['n_y'])
+
+            features, final_data, locs_table = _spectra_from_image_2d(intensities, wn,
+                                                                      xpositions[x_loc],
+                                                                      ypositions[y_loc])

```

Comment:
If it is an image (not a list of arbitrary positions), then yes, you should use `_spectra_from_image` not `_spectra_from_image_2d`. It has the repeat/tile stuff already. 

Then you can just return the result of that function directly.

Reply:
From the code, there's no way the current reader can support arbitrary lists! It generates the coordinates from `x_init` / `x_delta` / 'n_x`. Are you sure `.fsm` files can have arbitrary lists?

If you make one helper, you then have to send it a variable to tell it what the data is. Better to have tools that do one thing well as we currently do. Maybe the names are slightly confusing, but they are well described in the docstrings.